base/abstractarray.jl

```diff
@@ -904,9 +904,18 @@ end
 
 function copyto!(dest::AbstractArray, dstart::Integer, src)
     i = Int(dstart)
-    for x in src
-        dest[i] = x
-        i += 1
+    if haslength(src)
+        checkbounds(dest, i)
+        checkbounds(dest, i + length(src) - 1)
+        for x in src
```

Comment:
How about replace this with
```julia
I = eachindex(dest)[i]
@inbounds for x in src
    dest[I] = x
    I = nextind(dest, I)
end
```
This might be faster for `IndexCartesian` cases.

Reply:
And it seems reasonable to improve `copyto!(dest::AbstractArray, src)` (L893) in this PR.

Reply:
Using `eachindex(dest)[i]` doesn't seem to help, on things I tried. 

But I agree that the whole-array method just above has room for comparable improvement.

Reply:
I could see some difference in the following example:
```julia
using BenchmarkTools
f(dest, i, src) = begin
    I = eachindex(dest)[i]
    @inbounds for x in src
        dest[I] = x
        I = nextind(dest, I)
    end
end

g(dest, i, src) = begin
    @inbounds for x in src
        dest[i] = x
        i += 1
    end
end

a = view(randn(100,100),1:100,1:100)
@btime f($a, 555, $(i + 1 for i in 1:1000))    # 2.144 μs (0 allocations: 0 bytes)
@btime g($a, 555, $(i + 1 for i in 1:1000))   # 2.956 μs (0 allocations: 0 bytes)
```

Reply:
Just when I thought I'd convinced myself... these give me:
```
julia> @btime f($a, 555, $(i + 1 for i in 1:1000))
  min 2.718 μs, mean 2.759 μs (0 allocations)

julia> @btime g($a, 555, $(i + 1 for i in 1:1000))
  min 804.804 ns, mean 810.625 ns (0 allocations)
```

Reply:
I tried it with 1.7.0 and master, on win64... The result is consistent with what I post above.
How could this happened?

BTW, how to make `@btime` show `min` and `mean` at the same time?

Reply:
Strange. I get similar numbers on 1.7 + rosetta. Above is master (minus 9 days?) on M1 mac native. 

(And, you need this PR https://github.com/JuliaCI/BenchmarkTools.jl/pull/258 ...)

Reply:
On an ancient xeon:
```
julia> @btime f($a, 555, $(i + 1 for i in 1:1000))
  6.117 μs (0 allocations: 0 bytes)

julia> @btime g($a, 555, $(i + 1 for i in 1:1000)) 
  8.277 μs (0 allocations: 0 bytes)

julia> versioninfo()
Julia Version 1.8.0-DEV.1001
Commit 65a6bca661 (2021-11-17 01:50 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2603 v4 @ 1.70GHz
```
and 1.7
```
julia> @btime f($a, 555, $(i + 1 for i in 1:1000)) 
  min 6.334 μs, mean 6.364 μs (0 allocations)

julia> @btime g($a, 555, $(i + 1 for i in 1:1000)) 
  min 10.044 μs, mean 10.064 μs (0 allocations)

julia> versioninfo()
Julia Version 1.7.0-beta3.0
Commit e76c9dad42 (2021-07-07 08:12 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2603 v4 @ 1.70GHz
```

Reply:
As for 2-arg version on `M1 native, master`.
I think the problem is `firstindex` always return `1` if `ndims != 1`. Use `first(eachindex(dest))` should make things consistent.

On the other hand, just notice that `nextind(A, ind::Base.SCartesianIndex2)` is not defined, so `dest` can't be `reinterpret(reshape, args...)`...

Reply:
Ok, good point re firstindex. So that's just handling offsets right now.

On the xeon, "g" is still an improvement on `Base.copyto!`, even if slower than "f" there. Is that true on your computer too?

Reply:
Well that's true, at least we eliminated the `boundscheck` within the loop.

Reply:
Sure:
```
julia> @btime f_each($a)
  min 9.250 μs, mean 9.401 μs (0 allocations)
32.23877902877461

julia> @btime f_linear($a)
  min 9.250 μs, mean 9.401 μs (0 allocations)
32.23877902877461

julia> @btime f_each($b)
  min 9.166 μs, mean 9.315 μs (0 allocations)
22.794925499363792

julia> @btime f_linear($b)
  min 9.166 μs, mean 9.286 μs (0 allocations)
22.794925499363792
```

Reply:
On my machine, the result is 
```julia
@btime f_each($a) # 9.100 μs (0 allocations: 0 bytes)
@btime f_linear($a) # 22.700 μs (0 allocations: 0 bytes)
@btime f_each($b)  #  9.000 μs (0 allocations: 0 bytes)
@btime f_linear($b) # 22.500 μs (0 allocations: 0 bytes)
```
Is this some "dark technic" of M1?